### PR TITLE
feat: add `det deploy gke-experimental` [DET-5752]

### PR DESCRIPTION
**Pull request by ioga**
## Description

I've inherited #2842 and did a minimal amount of changed and testing to make it mergeable. Compared to the original PR,

- rename to `gke-experimental` to better highlight the nature of the tool.
- fixed a number of lint issues.
- added auto-naming for node pools and buckets (because providing these by hand is unnecessary and confusing).
- some extra flag validation to avoid basic failures.
- make `down` operation actually work.
- destroy GCS bucket on `down`.

## Test Plan

Basic:
1. `det deploy gke-experimental up --cluster-id <test-cluster-name>` (while in repo root, or pass `--helm-dir`)
2. kubectl get service determined-master-service-determined-gke -n default
3. export DET_MASTER=<ip address>:8080
4. Use it.
5. `det deploy gke-experimental down --cluster-id <test-cluster-name>`

There's a number of configuration flags. Use at your own risk.

NB: running `up` again on an existing cluster fails. there's no updates or upgrades.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/deploy/gke/cli.py
Comment (mackrorysd) [inline harness/determined/deploy/gke/cli.py:345]: This doesn't seem right that when referencing a CPU agent pool we're talking about a number of nodes required, and we're not showing how to set the priority of a task when telling people how to use the priority scheduler. Am I missing something?

File: harness/determined/deploy/gke/cli.py
Comment (mackrorysd) [inline harness/determined/deploy/gke/cli.py:156]: Would be nice to see if gsutil, gcloud, etc. is on the PATH and fail with a helpful message if it's missing.

File: harness/determined/deploy/gke/cli.py
Comment (mackrorysd) [inline harness/determined/deploy/gke/cli.py:509]: --multi-node-pool sounds like it's referring to a pool with >1 nodes, not multiple node pools. Would suggest --multiple-node-pools. I suppose the reason to not just do both like we do for other platforms is because auto-scaling can't work with our recommended scheduler?

File: harness/determined/deploy/gke/cli.py
Comment (mackrorysd) [inline harness/determined/deploy/gke/cli.py:579]: Should be 'no-managed-bucket' with the 'd', no?

File: harness/determined/deploy/gke/cli.py
Comment (ioga) [reply harness/determined/deploy/gke/cli.py:109]: This predates me but my understanding is that the goal was to allow configuring coscheduler on GPU pool only while having two pools.

Technically we have distributed CPU jobs, although I doubt there're any users on k8s using it.

File: harness/determined/deploy/gke/cli.py
Comment (ioga) [reply harness/determined/deploy/gke/cli.py]: IIRC if you've got 2 pools, and you want a coscheduler on CPUs, you've got to have it on GPUs as well or weird things happen.

File: harness/determined/deploy/gke/cli.py
Comment (ioga) [reply harness/determined/deploy/gke/cli.py:156]: I think it'd already give a somewhat understandable message: `FileNotFoundError: [Errno 2] No such file or directory: 'gsutil'`
Or do you want something better?

File: harness/determined/deploy/gke/cli.py
Comment (ioga) [reply harness/determined/deploy/gke/cli.py:345]: That really predates me. And honestly I haven't ever heard about CPU coscheduling setups working. I could cut this entire section if it's not useful. 
I'd say the original point of this tool was to avoid writing 8 command lines necessary to make a cluster. Not to try to make it produce all possible working configurations for everyone easily. I'd rather remove the guardrails to allow configurability.

File: harness/determined/deploy/gke/cli.py
Comment (mackrorysd) [reply harness/determined/deploy/gke/cli.py:156]: That works. Failing early before we've attempted ANY steps (and perhaps created a partial deployment) is a nice-to-have. It doesn't need to block the PR though.

File: harness/determined/deploy/gke/cli.py
Comment (mackrorysd) [reply harness/determined/deploy/gke/cli.py:109]: Ah - yeah I always interpret the "CPU training" work as meaning something other than what it means :smile:

File: harness/determined/deploy/gke/cli.py
Comment (mackrorysd) [reply harness/determined/deploy/gke/cli.py:345]: @eecsliu probably has the best context on this. Perhaps you could enlighten us on this issue and the ones I've mentioned above so we don't have to relearn anything the hard way? :smile:

Comment (eecsliu) [reply harness/determined/deploy/gke/cli.py:345]: yeah CPU training with coscheduling is broken because it requires a number of nodes, which you currently have to manually calculate. If you don't set this `pod-group.scheduling.sigs.k8s.io/min-available` field, the coscheduler will ignore it. However, it then becomes no better than the default scheduler.

File: harness/determined/deploy/gke/cli.py
Comment (eecsliu) [reply harness/determined/deploy/gke/cli.py]: A little late to this party, but schedulers in k8s are global, and not per resource pool. So while it is possible to run multiple schedulers in a cluster, the current state of determined (since k8s doesn't have resource pools yet) is that we can only easily use one type of scheduler at a time.

File: harness/determined/deploy/gke/cli.py
Comment (mackrorysd) [reply harness/determined/deploy/gke/cli.py:109]: The split is separate flags to enable the scheduler for the GPU / CPU resource pools.

Comment (eecsliu) [reply harness/determined/deploy/gke/cli.py:109]: hmm well it makes sense to check for whether the coscheduler and the preemption scheduler are both set (since only one should be set at a time), but it is not currently possible to run a CPU pool alongside a GPU pool.

Comment (ioga) [reply harness/determined/deploy/gke/cli.py:109]: I believe the case which is handled here is that you may have slotted jobs running on GPUs with coscheduler; and then there's CPU-only nodes without coscheduler for zero-slot tasks; and there's some amount of code to setup taints/tolerations here to ensure that stuff is getting scheduled that way.

If you also want coscheduler on CPU's, this code will also force you to have coscheduler on GPUs as well. So that case is covered.

I could rip it out if you think it doesn't make sense. 

But in this tool I'd try to lean towards being more flexible than having guardrails and only allowing safe and sound setups. This thing is not recommended for production really. I see it as a development tool. So that if you're testing something, you could setup a configuration using some combination of flags, and give that combination of flags to other people - which is better than having to do all this setup manually (where you've got to run 5+ pretty long command lines).

Comment (ioga) [reply harness/determined/deploy/gke/cli.py:109]: alright, I've ripped it out and rebased the whole branch. PTAL.

**Review by mackrorysd — APPROVED** [review]
Thanks for all the iteration on this. LGTM if we can get it to not fall over on Rehan's credit.

**Pull request merged**
